args <- commandArgs(trailingOnly = TRUE)
# the first input parameter needs to be the directory for metagenomics with all the result subdirectories
# the second input parameter needs to be the directory for metatranscriptomics with all the result subdirectories
# the third input parameter needs to be the directory where the results will be saved and the subdirectories created - needs to exist

# assumes a specific directory structure generated by default by the earlier scripts

# e.g.
# args <- character(3)
# args[1] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metagenomics"
# args[2] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metatranscriptomics"
# args[3] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/downstream/lmm_analysis"

# a script to perform linead mixed effects model (LMM) analysis for the metabolic marker genes related to nitrogen and methane

# load libraries
library(lmerTest)

# define nitrogen and methane related genes to be analyzed
nitrogen_genes <- c("NirS", "NirK", "NorB", "NosZ","NifH", "NapA", "NrfA", "NarG", "HzsA")
methane_genes <- c("PmoA", "McrA")

# load datas - analyze all datas similarly
datas_list <- list()

# metabolic marker gene datas
# metagenomics
setwd(as.character(args[1]))
setwd("metmarkdb_diamond")
load("Matrices_For_Downstream.RData")

# investigate the nitrogen and methane related genes

# nitrogen
data <- metmark_tpm_data
nit_genes <- nitrogen_genes[which(nitrogen_genes%in%rownames(data))]
nit_data <- data[nit_genes,]

# methane
meth_genes <- methane_genes[which(methane_genes%in%rownames(data))]
meth_data <- data[meth_genes,]

# combine into one
data <- rbind(nit_data, meth_data)

# add into list
datas_list[[1]] <- data

# metatranscriptomics
setwd(as.character(args[2]))
setwd("metmarkdb_diamond")
load("Matrices_For_Downstream.RData")

# nitrogen
data <- metmark_tpm_data
nit_genes <- nitrogen_genes[which(nitrogen_genes%in%rownames(data))]
nit_data <- data[nit_genes,]

# methane
meth_genes <- methane_genes[which(methane_genes%in%rownames(data))]
meth_data <- data[meth_genes,]

# combine into one
data <- rbind(nit_data, meth_data)

# add into list
datas_list[[2]] <- data

# name the datas
names(datas_list) <- c("metmark_mg", "metmark_mt")

# log2 transform the datas for the LMM models
for(i in 1:length(datas_list)){
  datas_list[[i]] <- log2(datas_list[[i]]+1)
}

# pick the relevant metadata
meta_use <- metadata[,c("Grazing","Treatment","Veg_clusters")]
colnames(meta_use) <- c("grazing", "treatment", "veg_clusters")

# set levels for metadata
meta_use$grazing <- factor(meta_use$grazing, levels = c("ungrazed", "grazed"))
meta_use$treatment <- factor(meta_use$treatment, levels = c("CTL", "+S", "-S"))
meta_use$veg_clusters <- factor(meta_use$veg_clusters, levels = c("t_ces", "c_cho", "c_ros"))

# set working directory to the results directory
setwd(as.character(args[3]))
keep_items <- ls()
keep_items <- c(keep_items, "keep_items")

for(i in 1:length(datas_list)){
  
  print(paste("Processing data:", names(datas_list)[i]))
  
  # pick the relevant data
  data <- datas_list[[i]]
  
  # save the values for the simple models here
  sig_values_lmm_simple <- coefficients_lmm_simple <- data.frame(matrix(nrow = nrow(data), ncol = 3))
  
  # save the values for the best fitting models here
  sig_values_lmm_best <- coefficients_lmm_best <- data.frame(matrix(nrow = nrow(data), ncol = 6))
  
  # save the models also
  lmm_models_simple <- list()
  lmm_models_complex <- list()
  
  for(k in 1:nrow(data)){
    
    # make a data frame
    temp_data <- cbind(as.numeric(data[k,]), meta_use)
    colnames(temp_data)[1] <- "expression"
    
    # build the simpler model - no interactions
    set.seed(1)
    mod_lmm_simple <- lmer(expression ~ grazing + treatment + (1|veg_clusters), REML=FALSE, data = temp_data)
    
    # build the more complex model including interactions
    set.seed(1)
    mod_lmm_complex <- lmer(expression ~ grazing*treatment + (1|veg_clusters), REML=FALSE, data = temp_data)
    
    # get the coefficients
    sum_lmm_simple <- summary(mod_lmm_simple)$coefficients
    sum_lmm_complex <- summary(mod_lmm_complex)$coefficients
    
    # perform Likelihood Ratio Test (LRT)
    # concept: The LRT compares the likelihoods of the two models. It's based on the idea that if the 
    # interaction term significantly improves the model's fit, the model with the interaction should 
    # have a much higher likelihood.
    model_comparison <- anova(mod_lmm_simple, mod_lmm_complex)
    
    # extract significance value
    model_comparison_sig <- model_comparison$`Pr(>Chisq)`[2]
    
    # save the signficance values for the simple model
    sig_values_lmm_simple[k,] <- sum_lmm_simple[-1,5]
    coefficients_lmm_simple[k,] <- sum_lmm_simple[-1,1]
    
    # save the significance values for the best model
    if(model_comparison_sig<0.05){
      sig_values_lmm_best[k,1:5] <- sum_lmm_complex[-1,5]
      sig_values_lmm_best[k,6] <- "interaction"
      
      coefficients_lmm_best[k,1:5] <- sum_lmm_complex[-1,1]
      coefficients_lmm_best[k,6] <- "interaction"
    }else{
      sig_values_lmm_best[k,1:3] <- sum_lmm_simple[-1,5]
      sig_values_lmm_best[k,6] <- "simple"
      
      coefficients_lmm_best[k,1:3] <- sum_lmm_simple[-1,1]
      coefficients_lmm_best[k,6] <- "simple"
    }
    
    # save the models
    lmm_models_simple[[k]] <- mod_lmm_simple
    lmm_models_complex[[k]] <- mod_lmm_complex
    
    # clean 
    rm(temp_data)
    rm(mod_lmm_simple)
    rm(mod_lmm_complex)
    rm(sum_lmm_simple)
    rm(sum_lmm_complex)
    rm(model_comparison)
    rm(model_comparison_sig)
  }
  
  # set names
  names(lmm_models_simple) <- names(lmm_models_complex) <- rownames(sig_values_lmm_simple) <- rownames(coefficients_lmm_simple) <- rownames(sig_values_lmm_best) <- rownames(coefficients_lmm_best) <- rownames(data)
  colnames(sig_values_lmm_best)[1:5] <- colnames(coefficients_lmm_best)[1:5] <- rownames(summary(lmm_models_complex[[1]])$coefficients)[-1]
  colnames(sig_values_lmm_best)[6] <- colnames(coefficients_lmm_best)[6] <- "best_model"
  colnames(sig_values_lmm_simple) <- colnames(coefficients_lmm_simple) <- rownames(summary(lmm_models_simple[[1]])$coefficients)[-1]
  
  # save everything
  save_items <- ls()
  save_items <- save_items[-which(save_items%in%keep_items)]
  
  dir.create(names(datas_list)[i])
  setwd(names(datas_list)[i])
  save(list = save_items, file = "lmm_analysis_done.RData")
  setwd("../")
  
  # clean
  save_items <- c(save_items, "save_items")
  rm(list = save_items)
  
  print("Data processed.")
  print("***************************************************************")
  
}

# print out session info
print("SessionInfo:")
sessionInfo()
