args <- commandArgs(trailingOnly = TRUE)
# the first input parameter needs to be the directory for metagenomics with all the result subdirectories
# the second input parameter needs to be the directory for metatranscriptomics with all the result subdirectories
# the third input parameter needs to be the directory where the results will be saved and the subdirectories created - needs to exist

# assumes a specific directory structure generated by default by the earlier scripts

# e.g.
# args <- character(3)
# args[1] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metagenomics"
# args[2] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metatranscriptomics"
# args[3] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/downstream/metabolic_functional_datas/lmm_analysis/kegg_gsea_enrichment"

# small script to export the KO groups in the KEGG datas for anvi'o anv-estimate-metabolism for module completeness estimation

# load KEGG datas
datas_list <- list()

# metagenomics
setwd(as.character(args[1]))
setwd("kegg_diamond")
load("Matrices_For_Downstream.RData")

# add into list
datas_list[[1]] <- ko_tpm_data

# metatranscriptomics
setwd(as.character(args[2]))
setwd("kegg_diamond")
load("Matrices_For_Downstream.RData")

# add into list
datas_list[[2]] <- ko_tpm_data

# name the datas
names(datas_list) <- c("kegg_mg", "kegg_mt")

# set working directory
setwd(as.character(args[3]))
keep_items <- ls()
keep_items <- c(keep_items, "keep_items")

for(i in 1:length(datas_list)){
  
  # pick the relevant data
  data <- datas_list[[i]]
  
  # get KO groups present in the chosen data for anvio-estimate-metabolism - parse into proper format into an enzyme file
  ko_enzymes <- cbind(paste("r", seq(1:nrow(data)), sep = ""), rownames(data), rep("KEGG", nrow(data)))
  colnames(ko_enzymes) <- c("gene_id",	"enzyme_accession",	"source")
  
  # save the KO enzymes in data
  dir.create(names(datas_list)[i])
  setwd(names(datas_list)[i])
  write.table(x = ko_enzymes, file = "kegg_enzymes_for_anvio.txt", quote = F, row.names = F, col.names = T, sep = "\t")
  setwd("../")
  
  # clean
  del_items <- ls()
  del_items <- del_items[-which(del_items%in%keep_items)]
  rm(list = del_items)
}

# print out session info
print("SessionInfo:")
sessionInfo()
