args <- commandArgs(trailingOnly = TRUE)
# the first input parameter needs to be the directory for metagenomics with all the result subdirectories
# the second input parameter needs to be the directory for metatranscriptomics with all the result subdirectories
# the third input parameter needs to be the directory where the results will be saved and the subdirectories created - needs to exist

# assumes a specific directory structure generated by default by the earlier scripts

# e.g.
# args <- character(3)
# args[1] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metagenomics"
# args[2] <- "/scratch/project_2009164/2_OULANKA/Tommi/final/metatranscriptomics"
# args[3] <- /scratch/project_2009164/2_OULANKA/Tommi/final/downstream/kegg_modules"

# small script to export the KO groups in the KEGG datas for anvi'o anv-estimate-metabolism for module completeness estimation

# load KEGG datas
datas_list <- list()

# metagenomics
setwd(as.character(args[1]))
setwd("kegg_diamond")
load("Matrices_For_Downstream.RData")

# add into list
datas_list[[1]] <- ko_tpm_data

# metatranscriptomics
setwd(as.character(args[2]))
setwd("kegg_diamond")
load("Matrices_For_Downstream.RData")

# add into list
datas_list[[2]] <- ko_tpm_data

# name the datas
names(datas_list) <- c("kegg_mg", "kegg_mt")

# set working directory
setwd(as.character(args[3]))
keep_items <- ls()
keep_items <- c(keep_items, "keep_items")

for(dat in 1:length(datas_list)){
  data <- datas_list[[dat]]
  
  # put into own directory
  dir.create(names(datas_list)[dat])
  setwd(names(datas_list)[dat])
  
  # go through the samples in the data separately and record the KO groups present in each sample
  for(sam in 1:ncol(data)){
    zero_vals <- which(data[,sam]==0)
    
    if(length(zero_vals)>0){
      temp_data <- data[-zero_vals,]
    }else{
      temp_data <- data
    }
    
    # save the KO enzymes
    # put each sample into own directory
    dir.create(colnames(data)[sam])
    setwd(colnames(data)[sam])
    
    # get KO groups present in the chosen data for anvio-estimate-metabolism - parse into proper format into an enzyme file
    ko_enzymes <- cbind(paste("r", seq(1:nrow(temp_data)), sep = ""), rownames(temp_data), rep("KEGG", nrow(temp_data)))
    colnames(ko_enzymes) <- c("gene_id",	"enzyme_accession",	"source")
    
    write.table(x = ko_enzymes, file ="kegg_enzymes_for_anvio.txt", quote = F, row.names = F, col.names = T, sep = "\t")
    setwd("../")
  }
  
  setwd("../")
  
  # clean
  del_items <- ls()
  del_items <- del_items[-which(del_items%in%keep_items)]
  rm(list = del_items)
}

# print out session info
print("SessionInfo:")
sessionInfo()
